Realm/Tests/KVOTests.mm

```diff
@@ -0,0 +1,1278 @@
+        XCTAssertEqualObjects(note[NSKeyValueChangeNewKey], NSNull.null);
+    }
+}
+
+- (void)testDeleteLinkedToObjectViaTableClear {
+    KVOLinkObject2 *obj = [self createLinkObject];
+    KVORecorder r(self, obj, @"obj");
+    [self.realm deleteObjects:[KVOLinkObject1 allObjectsInRealm:self.realm]];
+
+    if (NSDictionary *note = AssertNotification(r)) {
+        XCTAssertTrue([note[NSKeyValueChangeOldKey] isKindOfClass:[RLMObjectBase class]]);
+        XCTAssertEqualObjects(note[NSKeyValueChangeNewKey], NSNull.null);
+    }
+}
+
+- (void)testDeleteLinkedToObjectViaQueryClear {
+    KVOLinkObject2 *obj = [self createLinkObject];
+    KVORecorder r(self, obj, @"obj");
+    [self.realm deleteObjects:[KVOLinkObject1 objectsInRealm:self.realm where:@"TRUEPREDICATE"]];
+
+    if (NSDictionary *note = AssertNotification(r)) {
+        XCTAssertTrue([note[NSKeyValueChangeOldKey] isKindOfClass:[RLMObjectBase class]]);
+        XCTAssertEqualObjects(note[NSKeyValueChangeNewKey], NSNull.null);
+    }
+}
+
+- (void)testDeleteObjectInArray {
+    KVOLinkObject2 *obj = [self createLinkObject];
+    KVOLinkObject1 *linked = obj.obj;
+    [obj.array addObject:linked];
+    KVORecorder r(self, obj, @"array");
+    [self.realm deleteObject:linked];
+    AssertIndexChange(NSKeyValueChangeRemoval, [NSIndexSet indexSetWithIndex:0]);
+}
+
+- (void)testDeleteObjectsInArrayViaTableClear {
+    KVOLinkObject2 *obj = [self createLinkObject];
+    KVOLinkObject2 *obj2 = [self createLinkObject];
+    [obj.array addObject:obj.obj];
+    [obj.array addObject:obj.obj];
+    [obj.array addObject:obj2.obj];
+
+    KVORecorder r(self, obj, @"array");
+    [self.realm deleteObjects:[KVOLinkObject1 allObjectsInRealm:self.realm]];
+    AssertIndexChange(NSKeyValueChangeRemoval, ([NSIndexSet indexSetWithIndexesInRange:{0, 3}]));
+}
+
+- (void)testDeleteObjectsInArrayViaQueryClear {
+    KVOLinkObject2 *obj = [self createLinkObject];
+    KVOLinkObject2 *obj2 = [self createLinkObject];
+    [obj.array addObject:obj.obj];
+    [obj.array addObject:obj.obj];
+    [obj.array addObject:obj2.obj];
+
+    KVORecorder r(self, obj, @"array");
+    [self.realm deleteObjects:[KVOLinkObject1 objectsInRealm:self.realm where:@"TRUEPREDICATE"]];
+    AssertIndexChange(NSKeyValueChangeRemoval, ([NSIndexSet indexSetWithIndexesInRange:{0, 3}]));
+}
+
+@end
+
+// Mutate a different accessor backed by the same row as the accessor being observed
+@interface KVOMultipleAccessorsTests : KVOPersistedObjectTests
+@end
+@implementation KVOMultipleAccessorsTests
+- (id)observableForObject:(id)value {
+    if (RLMObject *obj = RLMDynamicCast<RLMObject>(value)) {
+        RLMObject *copy = [[obj.objectSchema.accessorClass alloc] initWithRealm:obj.realm schema:obj.objectSchema];
+        copy->_row = obj->_row;
+        return copy;
+    }
+    else if (RLMArray *array = RLMDynamicCast<RLMArray>(value)) {
+        return array;
```

Comment:
couldn't this, for a link view, return a new `RLMArrayLinkView` backed by the same linklist?